cms/djangoapps/contentstore/views/course_optimizer.py

```diff
@@ -0,0 +1,218 @@
+        Start a Celery task to check broken links in the course
+
+    The Studio UI uses a POST request to start the link check asynchronously, with
+    a link appearing on the page once it's ready.
+    """
+    course_key = CourseKey.from_string(course_key_string)
+    if not has_course_author_access(request.user, course_key):
+        raise PermissionDenied()
+    courselike_block = modulestore().get_course(course_key)
+    if courselike_block is None:
+        raise Http404
+    context = {
+        'context_course': courselike_block,
+        'courselike_home_url': reverse_course_url("course_handler", course_key),
+    }
+    context['status_url'] = reverse_course_url('link_check_status_handler', course_key)
+
+    # an _accept URL parameter will be preferred over HTTP_ACCEPT in the header.
+    requested_format = request.GET.get('_accept', request.META.get('HTTP_ACCEPT', 'text/html'))
+
+    if request.method == 'POST':
+        check_broken_links.delay(request.user.id, course_key_string, request.LANGUAGE_CODE)
+        return JsonResponse({'LinkCheckStatus': 1})
+    else:
+        # Only HTML request format is supported (no JSON).
+        return HttpResponse(status=406)
+
+
+@transaction.non_atomic_requests
+@require_GET
+@ensure_csrf_cookie
+@login_required
+@ensure_valid_course_key
+def link_check_status_handler(request, course_key_string):
+    """
+    Returns an integer corresponding to the status of a link check. These are:
+
+        -X : Link check unsuccessful due to some error with X as stage [0-3]
+        0 : No status info found (task not yet created)
+        1 : Scanning
+        2 : Verifying
+        3 : Success
+
+    If the link check was successful, an output result is also returned.
+    """
+    course_key = CourseKey.from_string(course_key_string)
+    if not has_course_author_access(request.user, course_key):
+        raise PermissionDenied()
+
+    # The task status record is authoritative once it's been created
+    task_status = _latest_task_status(request, course_key_string, link_check_status_handler)
+    json_content = None
+    test = None
+    response = None
+    error = None
+    broken_links_dto = None
+    if task_status is None:
+        # The task hasn't been initialized yet; did we store info in the session already?
+        try:
+            session_status = request.session["link_check_status"]
+            status = session_status[course_key_string]
+        except KeyError:
+            status = 0
+    elif task_status.state == UserTaskStatus.SUCCEEDED:
+        status = 3
+        artifact = UserTaskArtifact.objects.get(status=task_status, name='BrokenLinks')
+        with artifact.file as file:
+            content = file.read()
+            json_content = json.loads(content)
+            broken_links_dto = _create_dto(json_content, request.user)
+    elif task_status.state in (UserTaskStatus.FAILED, UserTaskStatus.CANCELED):
+        status = max(-(task_status.completed_steps + 1), -2)
```

Comment:
This is code I copied over from `import_export.py`. I believe the api will return a negative number on a fail related to the step in the process.
I agree this looks bad... but I would want to update code in both places at the same time. Maybe in a separate PR.

Reply:
I see people often just copying over code from older places since it just works, and thus creating new not-great code. I don't generally think that's a good idea. In my opinion we can extract code that's reused into helper functions if the code is good, but if not, I'd prefer writing new code that is better. It also means the code author needs to think and understand the code a little bit more in-depth than when they copy it.

If you want to change the code but it's too much trouble to extract the function from the other place, I'd say just change it in the new place and then just extract the parts that are still the same

Reply:
_Focus:_
Above the line - important